great_expectations/analytics/base_event.py

```diff
@@ -0,0 +1,82 @@
+from dataclasses import dataclass
+from typing import ClassVar, List, Optional
+from uuid import UUID
+
+from great_expectations.analytics.config import get_config
+
+
+@dataclass(frozen=True)
+class Action:
+    """Action is the "verb" representing what happened.
+
+    Attributes:
+        name: A description of what happened. For example (<object>.<verb>) "validation_result.saved" or "token.deleted"
+    """
+
+    name: str
+
+    def __repr__(self):
+        return self.name
+
+
+@dataclass
+class Event:
+    """Details of an analytics event.
+
+    Attributes:
+        action: The "verb" describing what action this event represents.
+    """
+
+    action: Action
+
+    @property
+    def data_context_id(self) -> UUID:
+        return get_config().data_context_id
+
+    @property
+    def organization_id(self) -> UUID:
+        return get_config().organization_id
+
+    @property
+    def oss_id(self) -> UUID:
+        return get_config().oss_id
+
+    @property
+    def user_id(self) -> UUID:
+        return get_config().user_id
+
+    _allowed_actions: ClassVar[Optional[List[Action]]] = None
+
+    def __post_init__(self):
+        allowed_actions = self.get_allowed_actions()
+        if (
+            allowed_actions is not None
```

Comment:
I feel like this protection isn't really protecting us from much; subclasses have to opt in to using it, and then the subclass defines both its action and that the action is allowed. Maybe there are cases I'm missing though.

Reply:
@tyler-hoffman in the case of this specific `Event`, there was only one `Action` that can be associated with it and to make using the event a bit easier, I set it in the ctor. However, there's another, more general pattern here. For example, a `CheckpointEvent` may support multiple actions, e.g. `checkpoint.created`, `checkpoint.updated`, `checkpoint.deleted`.

The use of `allowed_actions` become clearer once we add code to generate an event schema file for use in our analytics pipelines.